src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.StringSegment.cs

```diff
@@ -0,0 +1,496 @@
+            // So this method will get only <3-size prefix code unit><continuation>. Because we know more data will not be able
+            // to complete this code point, we need to pass isFinalSegment = true to ensure that the encoder consumes this data eagerly
+            // instead of leaving it and returning NeedsMoreData.
+            WriteStringSegmentEscape(partialCodePointBuffer, true);
+
+            WriteStringSegmentEscape(utf8Value, isFinalSegment);
+        }
+
+        private void WriteStringSegmentEscape(ReadOnlySpan<byte> utf8Value, bool isFinalSegment)
+        {
+            if (utf8Value.IsEmpty) return;
+
+            int escapeIdx = JsonWriterHelper.NeedsEscaping(utf8Value, _options.Encoder);
+            if (escapeIdx != -1)
+            {
+                WriteStringSegmentEscapeValue(utf8Value, escapeIdx, isFinalSegment);
+            }
+            else
+            {
+                WriteStringSegmentData(utf8Value);
+            }
+        }
+
+        private void WriteStringSegmentEscapeValue(ReadOnlySpan<byte> utf8Value, int firstEscapeIndexVal, bool isFinalSegment)
+        {
+            Debug.Assert(int.MaxValue / JsonConstants.MaxExpansionFactorWhileEscaping >= utf8Value.Length);
+            Debug.Assert(firstEscapeIndexVal >= 0 && firstEscapeIndexVal < utf8Value.Length);
+            byte[]? valueArray = null;
+            int length = JsonWriterHelper.GetMaxEscapedLength(utf8Value.Length, firstEscapeIndexVal);
+            Span<byte> escapedValue = length <= JsonConstants.StackallocByteThreshold ?
+                stackalloc byte[JsonConstants.StackallocByteThreshold] :
+                (valueArray = ArrayPool<byte>.Shared.Rent(length));
+
+            JsonWriterHelper.EscapeString(utf8Value, escapedValue, firstEscapeIndexVal, _options.Encoder, out int consumed, out int written, isFinalSegment);
+
+            WriteStringSegmentData(escapedValue.Slice(0, written));
+
+            Debug.Assert(consumed == utf8Value.Length || !isFinalSegment);
+            if (utf8Value.Length != consumed)
+            {
+                Debug.Assert(!isFinalSegment);
+                Debug.Assert(utf8Value.Length - consumed < 4);
+                SetPartialUtf8CodePoint(utf8Value.Slice(consumed));
+            }
+
+            if (valueArray != null)
+            {
+                ArrayPool<byte>.Shared.Return(valueArray);
+            }
+        }
+
+        private void WriteStringSegmentData(ReadOnlySpan<byte> escapedValue)
+        {
+            Debug.Assert(escapedValue.Length < int.MaxValue - 3);
+
+            int requiredBytes = escapedValue.Length;
+
+            if (_memory.Length - BytesPending < requiredBytes)
+            {
+                Grow(requiredBytes);
+            }
+
+            Span<byte> output = _memory.Span;
+
+            escapedValue.CopyTo(output.Slice(BytesPending));
+            BytesPending += escapedValue.Length;
+        }
+
+        private void WriteStringSegmentProlog()
```

Comment:
Unless you mean the programming language 😊
```suggestion
        private void WriteStringSegmentPrologue()
```